Update the contributing docs (#43931)
microsoft/TypeScript: CONTRIBUTING_CODE.md

## The TypeScript Compiler

There's a lot to learn, and this document is going to try and give you a very high level overview of how the
TypeScript compiler works and how you can contribute to it.

Here are the key sections in this docs:

- [Getting Set up](#getting-set-up)
- [What is in the TypeScript repo?](#what-is-in-the-typescript-repo)
- [The TypeScript Compiler](#the-typescript-compiler)
- [The TSServer](#the-tsserver)
- [Tips and Tricks](#tips-and-tricks)

## Getting set up

You need a modern version [of Node.js](https://nodejs.org/en/) installed and then:

- Clone the repo: `git clone https://github.com/microsoft/TypeScript` (optional: `--depth=1`)
- `cd TypeScript`
- `npm i` - installs dependencies

> - `npm run gulp` - builds the TypeScript compiler

Reviewer: I still run `gulp local`. Is that wrong?

Reviewer: I guess that's the default task .-.

- `code .` then briefly come back to terminal to run:
- `npm test` which will verify everything works fine

With that done, you should have a copy of the compiler up and running with tests running in the background to
prove that everything is working correctly. This could take around 10m, you should see be a progress indicator in
your terminal.

As an alternative, this repository also includes a [development container](https://code.visualstudio.com/docs/remote/containers) that you can use to quickly create an isolated development environment with all the tools you need to start working on TypeScript. To get started with a dev container and VS Code, either:

- Clone the TypeScript repository locally and use the `Open Folder in Container` command.

> - Use the `Clone Repository in Container Volume` command to clone the TypeScript repository into a new container.

Reviewer: ```suggestion
- Use the `Clone Repository in Container Volume` command to clone the TypeScript repository into a new container.
The development container uses a Linux image so you will need to switch to Linux containers if you are using Docker for Windows.
```

### Build commands

You can use `npm run gulp -- --tasks --depth 1` to see all the possible build commands.

### What happened during setup

> TypeScript is built in TypeScript. The compiler uses a "last known good" version of the TypeScript to bootstrap,

Reviewer: ```suggestion
TypeScript is written in and built by TypeScript. The compiler uses a "last known good" version of the TypeScript build to bootstrap.
```

> you can find this version in `/lib`. The command `npm run gulp` created a copy of the compiler at `built/local`.

Reviewer: ```suggestion
You can find this version in `/lib`. The command `npm run gulp` created a copy of the compiler at `built/local`.
```

The corresponding version of the `tsc` compiler used in the npm package is now available via:
`node built/local/tsc.js --help`.

TypeScript has a _comprehensive_ test suite, most of the tests are high level integration tests which you can find
at `tests/cases`. Given the scope and depth of a compiler like TypeScript, we use a lot of tests to ensure
improvements don't break old requirements.

### Configuring your editor

To get your VS Code environment working smoothly, set up the per-user config:

- Set up your `.vscode/launch.json` by running: `cp .vscode/launch.template.json .vscode/launch.json`
- Set up your `.vscode/settings.json` by running: `cp .vscode/settings.template.json .vscode/settings.json`

## What is in the TypeScript repo?

The TypeScript repo is a single codebase which creates a set of files which are released as the "typescript" npm
module:

- `typescript.js` - The compiler API
- `tsc.js` - The command line experience for using TypeScript, which uses `typescript.js`

> - `typescriptServices.js` - A web server which responds to questions about TS/JS code from editors, which uses

Reviewer: Curious - who uses this, and what do they use it for?

Reviewer 2: Yes

Reviewer 3: It is just the services layer, in fact, it's nearly identical to `typescript.js`, the only difference being it has a UMD wrapper for browser embedding (and exposes a global `ts` in such an environment).

that was approved by the TC39 committee.

#### Type checking

TypeScript's type system works by creating symbols when "something" is declared in a scope. That "something" could
be a variable, type, interface, function or more. A symbol has a set of flags that indicate what kind(s) of 
declaration(s) it's for. Symbols are then compared to each other during assignment and are used throughout the type system for
checking.

For example, this code:

```ts
function hello() {
  console.log("Hi");
}

hello;
```

creates one new symbol `hello` whose declaration is the function `hello`. The following flag is set:
`SymbolFlags.Function`. The compiler looks up this symbol whenever the name `hello` appears, like on the last line of the code. 
You can use [this playground plugin](https://www.typescriptlang.org/play?install-plugin=playground-ts-symbols) to see the
symbols generated for any TS/JS code.

Creating symbols is the responsibility of the binder, which is the first part of the type checking process.

Type checking in TypeScript happens in the file `checker.ts`, this is a 40k line function which does a huge amount

> of work. The checker works primarily by recursively diving into, known as "walking", the syntax tree, and making assertions about the

Reviewer: The type checker walking the syntax tree should be mentioned at the beginning of the section since the reader has just learned what the syntax tree is.

nodes on its way through.

> As a rough outline, using the code above, the following functions in the checker would be called:

Reviewer: ```suggestion
As a rough outline, using the syntax tree in [Parsing text](#parsing-text), the following functions in the checker would be called:
```

Reviewer: This suggestion just to help connect this to the parser section a bit more, but it's understandable as is.

```
checkSourceFileWorker (for the root SourceFile)
 - checkSourceElementWorker (for each Statement)
  -  checkFunctionDeclaration (for the Function Declaration)
    - checkBlock (for the function's body Block)
      ...
```

Effectively each syntax node has its own type checking function, and that's usually a good place to start if you
want to add or amend a diagnostic which TypeScript raises.

Common changes to the type checker are PRs which revise, fix or create new error diagnostics around code patterns.

#### Emit

The code in `emitter.ts` creates output `.js`, `.d.ts`, `.map` files from `.js` or `.ts` files.

> The emitter first uses a series of transformers to take the input file's syntax tree and convert that into a new

Reviewer: Is there a name for the pre-transformation syntax tree in a given TS version?

syntax tree which fits the constraints of the tsconfig. For example if your target was `es2015` then the
transformers for Class Fields, ESNext, ES2020, ES2019, ES2018, ES2017, ES2016, ES2015, and Generators would run.
You can use
[this playground plugin](https://www.typescriptlang.org/play?install-plugin=playground-transformer-timeline) to
see the individual transformations for any TS/JS code.

After the transforms are finished, the emitter recurses through the new syntax tree printing out nodes into a
new text file.

Common changes to the emitter are PRs which transform new TypeScript syntax, or new JavaScript features from TC39,
into JavaScript that works on older platforms.

## TSServer

The TSServer is responsible for providing information to text editors. The TSServer powers features like code

> completion, refactoring tools and jump to definition. The TSServer is similar to the language server protocol

Reviewer: ```suggestion
completion, refactoring tools, and jump to definition. The TSServer is similar to the language server protocol
```

but is older.

> 

Reviewer: ```suggestion

Common changes to the server generally include additions and fixes to things used by editor UI and commands, such as refactorings and completions.

```
